crates/control/src/behavior/intercept_ball.rs

```diff
@@ -0,0 +1,45 @@
+use nalgebra::{point, Point2, UnitComplex};
+use types::{
+    HeadMotion, LineSegment, MotionCommand, OrientationMode, PathSegment, PrimaryState, WorldState,
+};
+
+pub fn execute(world_state: &WorldState) -> Option<MotionCommand> {
+    match (
+        world_state.robot.primary_state,
+        world_state.ball,
+        world_state.robot.robot_to_field,
+    ) {
+        (PrimaryState::Playing, Some(ball), Some(robot_to_field)) => {
+            let ball_in_front_of_robot = ball.ball_in_ground.coords.norm() < 2.0
+                && ball.ball_in_ground.x > 0.0
+                && ball.ball_in_ground.y.abs() < 0.5;
+            let ball_moving_towards_robot = ball.ball_in_ground_velocity.x < -0.15;
+            let ball_moving_towards_own_half =
+                (robot_to_field * ball.ball_in_ground_velocity).x < -0.05;
+
+            if !(ball_in_front_of_robot
+                && ball_moving_towards_robot
+                && ball_moving_towards_own_half)
+            {
+                return None;
+            }
+
+            let intercept_position = ball.ball_in_ground.y
+                - ball.ball_in_ground.x * ball.ball_in_ground_velocity.y
+                    / ball.ball_in_ground_velocity.x;
+            Some(MotionCommand::Walk {
+                head: HeadMotion::LookAt {
+                    target: ball.ball_in_ground,
+                },
+                path: vec![PathSegment::LineSegment(LineSegment(
+                    Point2::origin(),
+                    point![0.0, intercept_position],
+                ))],
```

Comment:
How does this exactly work?

Reply:
We calculate the position where the ball intersects the y axis in ground coordinates.
This is based on the relative ball position and velocity.
Since at this point we know the ball is moving towards the robot (i.e. negative velocity along the x axis), we can calculate how far the ball will move along the y axis by dividing the x position by the velocity slope.
This additional distance along the y axis is then combined with the current y position of the robot to get the intersection point.

Do you have suggestions to make this code more readable?

Reply:
Makes sense. I think it's pretty easy to understand if you know what ..._in_ground means (which you can expect), but I think a picture of the vectors would really help, but that's not feasible.